boa_engine/src/property/mod.rs

```diff
@@ -674,29 +675,36 @@ impl From<&[u16]> for PropertyKey {
 impl From<JsString> for PropertyKey {
     #[inline]
     fn from(string: JsString) -> Self {
-        parse_u32_index(string.as_slice().iter().copied()).map_or(Self::String(string), Self::Index)
+        parse_u32_index(string.as_slice().iter().copied())
```

Comment:
I think it's better to make `parse_u32_index` return a `Option<NonMaxU32>`.

[here](https://github.com/boa-dev/boa/pull/3321/files#diff-d70662fbce40a1232f13bd61110a0252420431ec8ca66da3d07c3003b9e01a63R650-R652) we should return with max check because it can be greater or equal to u32 max. And in the else case we can return with `new_unchecked` since it will never be greater or equal to max because of the digit count check.